### ISSUE TITLE
The strategy of math libraries

### ISSUE THREAD

**Issue by bvssvni** [PistonDevelopers/piston#842]
Piston uses vecmath because it is a library specially designed for this usage, and was never intended to expand the scope beyond what it does now. I would like to know what you think of keeping stuff modular or putting everything in one library. Personally I don't think vecmath has the proper design to deal with much growth, so I would like to make new small libraries as needed instead.

**Comment by ruud-v-a** [comment]
First, let me be clear what I understand under _modular_. There may be two different meanings:
- A system of components that fit together, where you can pick the components you need and don’t use the components you do not need (e.g. the `System.*` assemblies of the .NET framework).
- A system where multiple components can fulfil the same role, where a component can be swapped in favour of another one without affecting the rest of the system (e.g. a backend-agnostic renderer like Ogre).

And _decoupled_:
- Components are _coupled_ if they are aware of each other’s existence (e.g. `Option` and `Result` in the standard library, `Option` has `ok_or` returning `Result` and `Result` has `ok` returning `Option`).
- Components are decoupled if they are not coupled (e.g. a function in crate A taking a `Sender<T>` can be sending data to a function in crate B taking a `Receiver<T>`, even if there are no dependencies between A and B). Decoupling is a great way to design modular systems (in the second sense).

For decoupled systems, I think you always need some kind of contract, a point where components can connect. In the example above this is a shared dependency, `Sender`/`Receiver` in the standard library. (By the way, a great talk about this is [Simple Made Easy](http://www.infoq.com/presentations/Simple-Made-Easy) by Rich Hickey.)

Regarding quaternion/vecmath, they are obviously coupled because quaternion depends on vecmath (it is a one-way dependency though). I think this is a good thing, precisely because you need some kind of shared contract (the definition of a vector). Decoupling them would complicate things unnecessarily. With a bit of repetition you can get a quaternion without vectors, but one of the most prominent use cases for quaternions is rotating vectors, so there needs to be a dependency anyway. This is not modular in the second sense: you cannot swap vecmath for an other library to use with quaternion. It is modular in the first sense: if you do not need quaternions, you can pick vecmath only and ignore quaternion.

Note that this is all orthogonal to how the crates are organised: it is the functions and structs that depend on each other. They might be in the same mod, the same crate, or different crates. So now the question is, how do we organise this into mods and crates?

I did not measure the impact, but I expect that the overhead of a quaternion struct in vecmath (if you do not use it) is less than the crate overhead if you do use quaternion. More importantly, there is the overhead of versioning and publishing the extra crate. And I think https://github.com/PistonDevelopers/vecmath/commit/7945dad3c05bebdfb436ab6982c60984ed6e28f9 (which seems to be written for https://github.com/PistonDevelopers/quaternion/commit/bdf947aced93dd1b6c9101c79378bd328b07c6d9 specifically) demonstrates that there is inconvenience in having separate crates (though you can argue that vecmath needed the functions anyway).

TL;DR Design is independent of organisation.

**Comment by bvssvni** [comment]
@ruud-v-a What happened is you interrupting my flow! Don't do that!

Read https://github.com/PistonDevelopers/piston/issues/835

I appreciate your concern about the math libraries, but perhaps this should be discussed after we have something working.

What is most important to me is to be able to work without interruptions on a daily basis, and your comment expressed concern about maintenance, which I care a lot about, but without containing enough information for me to respond properly. This is very provoking! When you want to start discussions, please open up a new issue, and make yourself clearer. Don't think of me as a person you have to defend yourself against, because *I am on your side*.

**Comment by bvssvni** [comment]
When it comes to vecmath/quaternion, we all know that dependencies might change over time. A dependency that isn't exposed directly means you can decouple them later if you want to. This is one motivation behind the design of vecmath, to make it easier for libraries to switch to something else when they need to.

Quaternions are used differently than vectors. They are particularly useful for rotations, so the library should contains functionality for that. vecmath doesn't contain methods tied to a specific usage. For example, piston-graphics adds more functions to its math module that wouldn't fit directly within vecmath. There are large amounts of coordinate systems and transformations used in a game engine, and I want separate smaller libraries instead of a large one.

https://github.com/PistonDevelopers/vecmath/commit/7945dad3c05bebdfb436ab6982c60984ed6e28f9 was because this feature that lacked in vecmath, and was discovered when working on the quaternion library.

Because there are different use cases, these two libraries doesn't fit in the same patterns, and that is why I think there should be separate libraries.

**Comment by bvssvni** [comment]
@ruud-v-a You're right about modular design can be done within the crate.

One difference is caused by the PR process. If person A1 works on crate B1, and person A2 works on crate B2, then they can merge their changes individually. If both work on the same crate, then they need to coordinate their changes.

Another difference is navigation. The larger crate, the longer time it takes to navigate the source code. Two smaller crates are faster to navigate independently than if they are put together.

When you add these dimension to a library, the term *modular* gets a different context. It might be another word better suited better for this purpose, but that's the problem of language in general, and a common source to misunderstanding, that words changes meaning by the context they are used in.

**Comment by bvssvni** [comment]
@ruud-v-a Thanks for the link to the talk! It was useful.

There is a talk by Don Norman showing examples when simplicity is confusing https://github.com/PistonDevelopers/piston/issues/628

**Comment by bvssvni** [comment]
Will keep the libraries separate, closing for now.
